File: app/src/main/java/io/connorwyatt/flashcards/adapters/FlashcardTestPagerAdapter.java
package io.connorwyatt.flashcards.adapters;

import android.app.Fragment;
import android.app.FragmentManager;
import android.os.Bundle;

import java.util.List;

import io.connorwyatt.flashcards.data.entities.Flashcard;
import io.connorwyatt.flashcards.fragments.FlashcardTestCardFragment;
import io.connorwyatt.flashcards.fragments.FlashcardTestSummaryFragment;

public class FlashcardTestPagerAdapter extends FixedFragmentStatePagerAdapter {
    private List<Flashcard> flashcards;

    public FlashcardTestPagerAdapter(FragmentManager fragmentManager, List<Flashcard> flashcards) {
        super(fragmentManager);

        this.flashcards = flashcards;
    }

    private Fragment getSummaryFragment() {
        return new FlashcardTestSummaryFragment();
    }

    @Override
    public int getCount() {
        return flashcards.size() + 1;
    }

    @Override
    public int getItemPosition(Object object) {
        return POSITION_NONE;
    }

    @Override
    public Fragment getItem(int position) {
        boolean isFlashcard = position < flashcards.size();

        if (isFlashcard) {
            return getFlashcardFragment(position);
        } else {
            return getSummaryFragment();
        }
    }

    @Override
    public String getFragmentTag(int position) {
        String tag;

        if (position < flashcards.size()) {
            tag = flashcards.get(position).getId().toString();
        } else {
            tag = "summary";
        }

        return tag;
    }

    public Flashcard removeItem(int position) {
        Flashcard removedFlashcard = flashcards.remove(position);
        notifyDataSetChanged();
        return removedFlashcard;
    }

    private Fragment getFlashcardFragment(int position) {
        Flashcard currentFlashcard = flashcards.get(position);

        FlashcardTestCardFragment fragment = new FlashcardTestCardFragment();

        Bundle arguments = new Bundle();
        arguments.putLong(FlashcardTestCardFragment.ARGUMENT_KEYS.ID, currentFlashcard.getId());
        arguments.putString(FlashcardTestCardFragment.ARGUMENT_KEYS.TITLE, currentFlashcard
                .getTitle());
        arguments.putString(FlashcardTestCardFragment.ARGUMENT_KEYS.TEXT, currentFlashcard
                .getText());
        fragment.setArguments(arguments);

        return fragment;
    }
}

File: app/src/main/java/io/connorwyatt/flashcards/adapters/FlashcardTestPagerAdapter.kt
/*
 * Copyright (c) 2016-2017 Connor Wyatt <connorwyatt1@gmail.com>.
 *
 * This file can not be copied and/or distributed without the express permission of Connor Wyatt.
 */

package io.connorwyatt.flashcards.adapters

import android.app.Fragment
import android.app.FragmentManager
import android.os.Bundle
import android.support.v4.view.PagerAdapter
import io.connorwyatt.flashcards.data.entities.Flashcard
import io.connorwyatt.flashcards.fragments.FlashcardTestCardFragment
import io.connorwyatt.flashcards.fragments.FlashcardTestSummaryFragment

class FlashcardTestPagerAdapter(fragmentManager: FragmentManager)
  : FixedFragmentStatePagerAdapter(fragmentManager) {
  private val summaryFragment = FlashcardTestSummaryFragment()
  private var flashcards: MutableList<Flashcard>? = null

  //region Pager

  override fun getItem(position: Int): Fragment {
    val flashcard = flashcards?.getOrNull(position)

    return flashcard?.let { getFlashcardFragment(it) } ?: summaryFragment
  }

  override fun getItemPosition(item: Any?): Int {
    return PagerAdapter.POSITION_NONE
  }

  override fun getFragmentTag(position: Int): String {
    return if (position < flashcards?.size ?: 0) flashcards?.get(position)?.id!! else "summary"
  }

  override fun getCount(): Int {
    return (flashcards?.size ?: 0) + 1
  }

  //endregion

  //region Data

  fun getFlashcardById(id: String): Flashcard {
    return flashcards!!.find { it.id!! == id }!!
  }

  fun setData(flashcards: List<Flashcard>): Unit {
    this.flashcards = flashcards.toMutableList()
    notifyDataSetChanged()
  }

  fun removeItem(position: Int): Flashcard {
    val removedFlashcard = flashcards!!.removeAt(position)
    notifyDataSetChanged()
    return removedFlashcard
  }

  //endregion

  //region Fragments

  private fun getFlashcardFragment(flashcard: Flashcard): Fragment {
    val fragment = FlashcardTestCardFragment()

    val arguments = Bundle()

    arguments.putString(
      FlashcardTestCardFragment.Companion.ArgumentKeys.FLASHCARD_ID, flashcard.id!!)
    fragment.arguments = arguments


    return fragment
  }

  //endregion
}
